File: JS/Tree/107.Binary Tree Level Order Traversal II.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrderBottom = function(root) {
    const stack = [root];
    const result = [];
    let count = 1;
    let current = 0;
    let depth = 0;
    while(stack.length) {
        const node = stack.shift();
        if (node) {
            if (current === 0) {
                result.unshift([]);
            }
            result[0].push(node.val);
            current++;
            if (node.left) {
                stack.push(node.left);
            }
            if (node.right) {
                stack.push(node.right);
            }
            if (current === count) {
                current = 0;
                count = stack.length;
                
            }
        }
    }
    return result;
};

File: Golang/Tree/107.binary-tree-level-order-traversal-II.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func levelOrderBottom(root *TreeNode) [][]int {
	ret := [][]int{}
	if root != nil {
		stack := []*TreeNode{root}
		layer := 1
		for len(stack) > 0 {
			children := []*TreeNode{}
			temp := []int{}
			for i := 0; i < layer; i++ {
				node := stack[i]
				temp = append(temp, node.Val)
				if node.Left != nil {
					children = append(children, node.Left)
				}
				if node.Right != nil {
					children = append(children, node.Right)
				}
			}
			ret = append(ret, temp)
			layer = len(children)
			stack = children
		}
	}
	for i, j := 0, len(ret)-1; i <= j; {
		ret[i], ret[j] = ret[j], ret[i]
		i++
		j--
	}
	return ret
}
